api/api-notification-filter-v2/src/main/java/com/thoughtworks/go/apiv2/notificationfilter/NotificationFilterControllerV2.java

```diff
@@ -0,0 +1,202 @@
+            patch(Routes.NotificationFilterAPI.ID, mimeType, this::updateFilter);
+            delete(Routes.NotificationFilterAPI.ID, mimeType, this::deleteFilter);
+        });
+    }
+
+    public String index(Request request, Response response) throws IOException {
+        User user = userService.findUserByName(currentUsernameString());
+        List<NotificationFilter> notificationFilters = user.getNotificationFilters();
+
+        return writerForTopLevelObject(request, response, writer -> NotificationFiltersRepresenter.toJSON(writer, notificationFilters));
+    }
+
+    public String show(Request request, Response response) throws IOException {
+        long id = getIdFromRequest(request.params("id"));
+        User user = userService.findUserByName(currentUsernameString());
+        Optional<NotificationFilter> optionalFilter = user.getNotificationFilters().stream()
+                .filter(filter -> filter.getId() == id)
+                .findFirst();
+
+        if (optionalFilter.isEmpty()) {
+            throw new RecordNotFoundException(EntityType.NotificationFilter, id);
+        }
+
+        return writerForTopLevelObject(request, response, writer -> NotificationFilterRepresenter.toJSON(writer, optionalFilter.get()));
+    }
+
+    public String createNotificationFilter(Request request, Response response) throws IOException {
+        NotificationFilter notificationFilter = buildEntityFromRequestBody(request);
+        userService.addNotificationFilter(currentUserId(request), notificationFilter);
+        setEtagHeader(response, etagFor(notificationFilter));
+
+        return writerForTopLevelObject(request, response, writer -> NotificationFilterRepresenter.toJSON(writer, notificationFilter));
+    }
+
+    public String updateFilter(Request request, Response response) throws IOException {
+        long id = getIdFromRequest(request.params("id"));
+        NotificationFilter notificationFilter = buildEntityFromRequestBody(request);
+        User user = userService.findUserByName(currentUsernameString());
+
+        Optional<NotificationFilter> optionalFilter = user.getNotificationFilters().stream()
+                .filter(filter -> filter.getId() == id)
+                .findFirst();
+
+        if (optionalFilter.isEmpty()) {
+            throw new RecordNotFoundException(EntityType.NotificationFilter, id);
+        }
+
+        try {
+            notificationFilter.setId(id);
+            user.getNotificationFilters().remove(optionalFilter.get());
+            user.addNotificationFilter(notificationFilter);
+            userService.saveOrUpdate(user);
+            setEtagHeader(response, etagFor(notificationFilter));
+            return writerForTopLevelObject(request, response, writer -> NotificationFilterRepresenter.toJSON(writer, notificationFilter));
+        } catch (ValidationException e) {
+            throw new UnprocessableEntityException(e.getMessage());
+        }
```

Comment:
Should we move the updation logic to service same as create and update? It will make it easier to test.

Reply:
Done